commit 5942b62db420781f3598ccc35b560bf66dc39965
Author: Matthew Hague <matthewhague@zoho.com>
Date:   Tue Sep 1 21:43:48 2015 +0100

    improved parsing times of large vcards

diff --git a/src/vc_parse.y b/src/vc_parse.y
index 61c36ad..b628b6b 100644
--- a/src/vc_parse.y
+++ b/src/vc_parse.y
@@ -33,6 +33,9 @@ extern FILE *yyin;
 extern int yylex ();
 void yyerror (char *s);
 
+extern long chars_in_buffer();
+extern void flush_buffer();
+
 vc_component *current_vcard = NULL;
 vc_component *current_vc = NULL;
 char *current_vc_param_name = NULL;
@@ -164,6 +167,9 @@ parse_vcard_file (FILE * fp)
     {
       vc = current_vcard;
     }
+  
+  fseek(yyin, -chars_in_buffer(), SEEK_CUR);
+  flush_buffer();
 
   return vc;
 }
diff --git a/src/vc_scan.l b/src/vc_scan.l
index e726c44..6ea02dd 100644
--- a/src/vc_scan.l
+++ b/src/vc_scan.l
@@ -23,14 +23,20 @@
 #define YYSTYPE char*
 #include "vc_parse.h"
 
-/****************************************************
-  to force flex to scan only one character at a time
- ****************************************************/
-#define YY_INPUT(buf,result,max_size) \
-    { \
-    int c = getc(yyin); \
-    result = (c == EOF) ? YY_NULL : (buf[0] = c, 1); \
-    }
+/**
+ * Two functions to enable only buffering one card at a time
+ */
+
+int chars_in_buffer()
+  {
+    return YY_CURRENT_BUFFER->yy_n_chars -
+           (yytext - YY_CURRENT_BUFFER->yy_ch_buf);
+  }
+
+void flush_buffer()
+  {
+    YY_FLUSH_BUFFER;
+  }
 
 /* TODO: clean up the token definitions to match the rfc */
 %}
